### ISSUE TITLE
Attempt to index a nil value - old_handler

### ISSUE THREAD

**Issue by asmodeus812** [j-hui/fidget.nvim#78]
Hello, due to a recent change in the following commit : https://github.com/j-hui/fidget.nvim/commit/e87666ce9420da887eff4e56430317a0acd8ad5e
I am receiving the following error, looking at the changes in the commit, the if statement makes no sense in my opinion and needs to be revisited. @ratmice
![image](https://user-images.githubusercontent.com/15955811/167639517-6119eb4e-7003-459f-9eb1-f6b3f132962f.png)
Using this to hook into jdtls by settin a custom progress handler. It was working perfectly fine until i pulled today. (I was on an older commit than the one mentioned above)
```Lua
local function get_progress(_, result, ctx)
    local lsp = vim.lsp
    local info = {
        client_id = ctx.client_id,
    }

    local kind = "report"
    if result.complete then
        kind = "end"
    elseif result.workDone == 0 then
        kind = "begin"
    elseif result.workDone > 0 and result.workDone < result.totalWork then
        kind = "report"
    else
        kind = "end"
    end

    local percentage = 0
    if result.totalWork > 0 and result.workDone >= 0 then
        percentage = result.workDone / result.totalWork * 100
    end

    local msg = {
        token = result.id,
        value = {
            kind = kind,
            percentage = percentage,
            title = result.task,
            message = result.task,
        },
    }
    lsp.handlers["$/progress"](nil, msg, info)
end
```

**Comment by j-hui** [comment]
The error you're seeing is from here: https://github.com/neovim/neovim/blob/v0.7.0/runtime/lua/vim/lsp/handlers.lua#L45

I'm not familiar with jdtls as I don't use it myself. Is it possible for jdtls to report progress without starting at workDone at 0%? If so, `client.messages.progress[token]` would never have been initialized, leading `client.messages.progress[token].message` to throw an error.

Can you see if you still encounter this error if you still use your handler, but _don't_ setup Fidget?

**Comment by ratmice** [comment]
I definitely think we're going to need more information, I have been running this patch alone and with lualine-lsp-progress,
If I understand the backtrace:

  1. your handler for jdtls calls fidgets progress handler
  2. fidget calls the builtin lsp progress handler, which crashes?

looking at jdtls I see this line here:
https://github.com/mfussenegger/nvim-jdtls/blob/master/lua/jdtls/setup.lua#L95

`vim.cmd("doautocmd <nomodeline> User LspProgressUpdate")`

In their progress handler, I wonder if that essentially is handling progress, clearing the `client.messages.progress` and then when the handler get calls again it is no longer initialized.

**Comment by ratmice** [comment]
Another thing you could try is keeping your setup the same, and then setting `vim.lsp.handlers["$/progress"] = nil` before setting up fidget, This should negate the effects of the patch entirely.

Anyhow my guess is that jdtls is mutating the `client.messages.progress` in a way which causes the builtin progress handler to crash.  Which wasn't a problem before since it was never called or perhaps only called once, but now that fidget will call it, it is either being callaed, or called twice but now in a state it isn't prepared for.

**Comment by ratmice** [comment]
So, I was able to reproduce this with the following slightly modified config:

```
require("fidget").setup{}
fidget_handler = vim.lsp.handlers["$/progress"]
local function get_progress(_, result, ctx)
  local lsp = vim.lsp
  local info = {
      client_id = ctx.client_id,
      id = ctx.client_id,
  }

  local kind = "report"
  if result.complete then
      kind = "end"
  elseif result.workDone == 0 then
      kind = "begin"
  elseif result.workDone > 0 and result.workDone < result.totalWork then
      kind = "report"
  else
      kind = "end"
  end

  local percentage = 0
  if result.totalWork > 0 and result.workDone >= 0 then
      percentage = result.workDone / result.totalWork * 100
  end

  local msg = {
      token = result.id,
      value = {
          kind = kind,
          percentage = percentage,
          title = result.task,
          message = result.task,
      },
  }
  -- lsp.handlers["$/progress"](nil, msg, info)
  fidget_handler(nil, msg, info)
end

vim.lsp.handlers["$/progress"] = function (_, result, ctx)
-- I stuffed this in a progress handler merely to grab a valid ctx.
-- This should crash on the same line when some lsp sends a progress.
-- We ignore the result pasesd in and replace it with the broken one.
get_progress(nil, {complete = false, workDone = 1, totalWork = 10, id = "umm", task = "foo"}, ctx)
end
```

A couple of things we can surmise from your code:
you are calling `vim.lsp.handlers["$/progress"]` from another handler (which is not `"$/progress"`, otherwise it would infinite loop).
you never add the tasks to `client.messages.progress` before calling the handler, the *builtin* progress handler (which) expects them to be there.

Once you change this to: it will work as before
```
vim.lsp.handlers["$/progress"] = nil
require("fidget").setup{}
```

IMO, the `get_progress` function here and whatever handler is calling it, just isn't setting up `client.progress.messages` to the extent that the builtin handlers expect the environment to be in when they when they are called from a normal progress message.

**Comment by j-hui** [comment]
@ratmice thanks for looking into this. As I already mentioned above, I'm still pretty sure that the problem is that whatever is producing progress reports is not creating `begin` messages.

In fact, this is made explicit in your code:

```lua
get_progress(nil, {complete = false, workDone = 1, totalWork = 10, id = "umm", task = "foo"}, ctx)
```

For this "progress message", you start with `workDone` of 1 (begin reports are only produced when `workDone = 0`). If I change that to 0, I no longer see the bug.

**Comment by j-hui** [comment]
> I don't think we should consider this a fidget bug since we can trigger the exact thing without fidget, I would think it must be either in the builtin handler, or your get_progress function.

**Comment by j-hui** [comment]
@asmodeus812 try adding this to your handler, before you call the progress handler:

```lua
    local client = vim.lsp.get_client_by_id(info.client_id)
    if client then
      client.messages.progress[msg.token] = client.messages.progress[msg.token] or {}
    end
    lsp.handlers["$/progress"](nil, msg, info)
```

Does that fix your issue?

**Comment by asmodeus812** [comment]
@j-hui not really, same issue is observed. The issue here is that for me at least, the possible workarounds both of you discussed above seem incredibly confusing. Nothing is really mentioned as far as possible configuration of the lsp progress handler in the docs, so i assume nothing needs to be done and fidget hooks in to that automatically. I am not sure what and how fidget handles the progress handler, and what it does with it. But i was expecting that simply calling the build in one from lsp will be enough, so fidget can figure out the rest - in other words, work **transparently** to the user. 
The same way vim.notify and the notify plugin works when it is hooked into vim.notify (crude example, i know, but i expect it should work similarly)

**Comment by j-hui** [comment]
No more messing around with custom handlers; closed by #143
